drivers/bridge/bridge.go

```diff
@@ -1340,6 +1341,81 @@ func (d *driver) DiscoverDelete(dType discoverapi.DiscoveryType, data interface{
 	return nil
 }
 
+func (d *driver) Restore(nid, eid string, sboxKey string, ifInfo driverapi.InterfaceInfo, options map[string]interface{}) error {
+	// restore endpoint
+	if ifInfo == nil {
+		return errors.New("invalid interface info passed")
+	}
+
+	// Get the network handler and make sure it exists
+	d.Lock()
+	n, ok := d.networks[nid]
+	d.Unlock()
+
+	if !ok {
+		return types.NotFoundErrorf("network %s does not exist", nid)
+	}
+	if n == nil {
+		return driverapi.ErrNoNetwork(nid)
+	}
+
+	// Sanity check
+	n.Lock()
+	if n.id != nid {
+		n.Unlock()
+		return InvalidNetworkIDError(nid)
+	}
+	n.Unlock()
+
+	// Check if endpoint id is good and retrieve correspondent endpoint
+	ep, err := n.getEndpoint(eid)
+	if err != nil {
+		return err
+	}
+
+	// Endpoint with that id exists either on desired or other sandbox
+	if ep != nil {
+		return driverapi.ErrEndpointExists(eid)
+	}
+
+	// Try to convert the options to endpoint configuration
+	epConfig, err := parseEndpointOptions(options)
+	if err != nil {
+		return err
+	}
+	endpoint := &bridgeEndpoint{id: eid, config: epConfig}
+	endpoint.macAddress = ifInfo.MacAddress()
+	endpoint.addr = ifInfo.Address()
+	endpoint.addrv6 = ifInfo.AddressIPv6()
+	endpoint.srcName = ifInfo.SrcName()
+
+	endpoint.containerConfig, err = parseContainerOptions(options)
+	if err != nil {
+		return err
+	}
+	endpoint.extConnConfig, err = parseConnectivityOptions(options)
+	if err != nil {
+		return err
+	}
+
+	if len(endpoint.extConnConfig.PortBindings) > 0 {
+		// TODO: daemon pass the ports to restore the portMapping may be a better way
+		endpoint.portMapping, err = n.allocatePorts(endpoint, n.config.DefaultBindingIP, d.config.EnableUserlandProxy)
```

Comment:
@aboch  thanks for your response
> Won't this will cause a new port mapping being done for each exposed port, a new instance of docker-proxy be spawned when there is already one running ?


docker-proxy will die once the daemon process dies, because we send SIGTERM to it if the daemon process dies
see https://github.com/docker/docker/blob/master/vendor/src/github.com/docker/libnetwork/portmapper/proxy.go#L110   so we have to spawned a new one. and 
> Also, if the user specified an explicit port mapping (-p X:x) for the container, won't this call fail when proxy tries to bind to an already used host port ?

also don't happened.  

>  Also, the old mapped ports are lost. If container is stopped, nobody will release the ports mapped during the previous daemon life.
I am afraid we cannot avoid storing the driver's endpoint to store.
Also, given portmapper is not storing the ports to store, he could give out already reserved ports to new containers (I know portmapper does not give out port that he has just freed, he keeps moving to the next one, but we should not relay on that) while designing the daemon reload changes.

The old mapped ports are not lost, the `n.allocatePorts(endpoint, n.config.DefaultBindingIP, d.config.EnableUserlandProxy)`  will re-allocate the ports of the container  and store these information in `PortMapper` , and to avoid duplication iptables rules, we add a check (https://github.com/docker/libnetwork/pull/1135/files/88e1c42a523211f8aba604be3373d98bac96b7c4#diff-b90cadcd0928c1e490272f4761a52bacR350) before insert. so `n.allocatePorts` here is to re-construct the `portMapper`. 
The problem is that the order of the sandboxes to be restored is changed(for example, A container start first, and then B, but during restore, the B may first), so for the random ports, the number could be changed, may not be the same with previous. 
But I still think we can avoid storing the driver's endpoint to store, we can pass the port mapping information form daemon( which we can see in docker ps), and use these information to re-construct `portMapper`, I working on this local, I'm sure it can work, I'll update this PR when I finished.